test/app-luatest/system_fiber_test.lua

```diff
@@ -0,0 +1,41 @@
+local server = require('test.luatest_helpers.server')
+local g = require('luatest').group()
+
+g.before_all = function()
+    g.server = server:new{alias = 'default'}
+    g.server:start()
+end
+
+g.after_all = function()
+    g.server:drop()
+end
+
+g.test_system_fibers = function()
+    g.server:exec(function(uri)
+        local fiber = require('fiber')
+        local t = require('luatest')
+
+        local conn = require('net.box').connect(uri)
+        local is_system = function(name)
+            return name:endswith('(net.box)') or
+                   name == 'gc' or name == 'checkpoint_daemon' or
+                   name:startswith('vinyl.')
+        end
+
+        local system_fibers = {}
+        for fid, f in pairs(fiber.info()) do
+            if is_system(f.name) then
+                -- fiber.find() + fiber:cancel() == fiber.kill()
+                local fiber_temp = fiber.find(fid)
+                table.insert(system_fibers, fiber_temp)
+                fiber_temp:cancel()
+            end
+        end
+
+        fiber.yield()
+        t.assert(conn:ping())
+        for _, f in pairs(system_fibers) do
+            t.assert(f:status() ~= 'dead')
```

Comment:
It is not really enough to protect only the Lua land. As I was saying, I can get to almost any system fiber via C API. And I don't need `fiber.find()` for this. System fibers often execute triggers where I can get `fiber_self()` and do with it anything:
```Lua
lib = box.lib.load('test')
kill_self = lib:load('oh_yes_kill_me')

net = require('net.box')
box.cfg{listen = 3313}
box.schema.user.grant('guest', 'super')
c = net.connect(box.cfg.listen, {reconnect_after = 1})
c:ping()

c:eval([[box.session.push(nil)]], {}, {on_push = function()
    kill_self()
end})
```
```C
#include "module.h"

int
oh_yes_kill_me(box_function_ctx_t *ctx, const char *args, const char *args_end)
{
	(void)ctx, (void)args, (void)args_end;
	fiber_cancel(fiber_self());
	return 0;
}
```
Compilation on Mac. `test.c` is stored in `tarantool/sandbox` with the C code above.
```
clang -dynamiclib -undefined dynamic_lookup -I../src/ -I../third_party/luajit/src/ -I../src/lib/msgpuck/ test.c -o test.dylib
```
Then copy the `test.dylib` into tarantool's workdir. This example is a bit exaggerated, but killing usually happens accidentally via more creative logic than here.

Reply:
We cannot delete `fiber_cancel` from the fiber's public C API. But can we change its API?

If so we can make `fiber_cancel` a wrapper around the function, the content of which will be equivalent to the current `fiber_cancel` (don't really know how to name it)). In this wrapper we can check if the fiber is system. If so we will set `diag` and return -1. We won't export internal function to public API. So, the fiber won't be ever cancelled from public. In tarantool itself we gonna use only this internal function. But we need to change the return value of `fiber_cancel`.  

@Gerold103 what do you think about this?

Reply:
Since nobody complained so far, I'd ignore this issue for now. We can always fix it later if we need to.

Reply:
Considering the fact, that the release is pretty close, we should merge something anyway. Overwise, there will be one more release with `error_reconnect` bug, which is actually solved by this PR. 

Now the fiber can be cancelled with sophisticated workarounds.  The possible solution: 

Let's make public C API function `fiber_cancel()` a wrapper, which will check if the fiber is system:
```C
void
fiber_cancel(struct fiber *f)
{
	/* Do nothing if the FIBER_IS_SYSTEM flag is set. */
	if ((f->flags & FIBER_IS_SYSTEM) != 0)
		return;

	fiber_cancel_system(f);
}
```
The content of the `fiber_cancel_system` is the same as current `fiber_cancel`.

We cannot change the API of the `fiber_cancel`, so let's just silently ignore cancellation request for a system fiber. This behavior is similar to the current one, but all checks will be done in public C API, and both API will behave the same way, as @Totktonada proposed. Moreover, it will be impossible to cancel the fiber from anywhere, as @Gerold103 proposed.

I created another PR for that solution: #7524. If it's all right or some minor changes are needed we can merge it. Overwise, 
let's merge this one, create ticket for cancelling system fibers from C and I will fix it ASAP after release.

Reply:
`fiber_cancel()` always ignores system fibers, we just had no system fibers before. If I'll say it this way, it will not be incompatible change.

If you're bother about API breakage, I completely missed point of this PR. Why this 'break' of C API is not okay, but the same is okay for the Lua API, which is more frequently used?

I requested a reason of the difference between C's and Lua's fiber cancel behavior (in a discussion above, not answered) and if your answer is 'backward compatibility', the answer is wrong. Technically speaking, cancelling of fibers that goes to be marked as system ones never produces a useful result. See compatibility terms definition in the [new release policy](https://github.com/tarantool/tarantool/discussions/6182) document:

> A code that are run on an older release is run with the same effect on a
newer release. However only meaningful code counts: if any code just give an
error, but starts doing something useful, the change is considered
compatible.

If you're bother about error reporting in `fiber_cancel()` (return value, box error area), just don't report an error and silently do nothing for a system fiber. We can add `fiber_cancel_v2()` later if there will be a demand.

Reply:
* `fiber_cancel()` always ignored "system" fibers. Except that the "systemness" was defined per-scope and using `FIBER_IS_CANCELLABLE` which we are trying to resurrect here in a better form;
* We don't need wrappers around it. Just make it not touch the system fibers. For forceful internal killing we need a new function with a good name (not `fiber_cancel_internal()`). Maybe `fiber_kill_sys()` or just `fiber_kill()` (a bit conflicting with Lua) or `fiber_cancel_forced()` or likewise. It shouldn't be exported to the public API in any way.
* About one more release with `error_reconnect` bug - I see no problem with that. The moment it appeared in at least one release we've already got it public. Doesn't matter now in how many releases. At least in version-flexible modules like vshard - they will need to workaround this bug anyway, whether you fix it now or next time.

As I was saying, I can't see much harm in the current form of this PR, but can't see much usefulness too. I can't vote it up, but Vova's ack should be enough anyway if you decide to move forward with this partial fix.

Reply:
`fiber_cancel_system` (matches `fiber_new_system`) sounds fine by me, but I'd be fine with doing this some time later, in a separate PR, as I said above. Anyway, I can't merge this PR unitl @Gerold103 approves:

<details>

![image](https://user-images.githubusercontent.com/1114511/182805926-431e7cc6-5791-4118-958f-0a9edcaeb709.png)

</details>

Reply:
Let's not merge this PR before the full solution is done. Current one is not full and the behavior of Lua's and C's `fiber_cancel` differ, which is not too great. 

----

> fiber_cancel() always ignored "system" fibers. Except that the "systemness" was defined per-scope and using FIBER_IS_CANCELLABLE which we are trying to resurrect here in a better form;

Didn't get your statement. `FIBER_IS_CANCELLABLE` is used only in order not to wakeup the fiber after its cancelling. However, currently all fibers can be cancelled if they call `fiber_testcancel()` inside. AFAICS, we don't want to set the flag `FIBER_IS_CANCELLED` in case of system fiber at all. 

----

> We don't need wrappers around it.

Why not? The logic of the `fiber_cancel` should remain the same, excepting checking for systemness. `fiber_cancel_system` is supposed to work the same way the `fiber_cancel` currently works. If I'm not right, could you please explain, why and what will be the differences in behavior of the current `fiber_cancel` and `fiber_cancel_system`. 

----

> About one more release with error_reconnect bug - I see no problem with that.

Got you, lets postpone the merging.

Reply:
> Didn't get your statement. FIBER_IS_CANCELLABLE is used only in order not to wakeup the fiber after its cancelling

В теории это так. По факту его использовали где ни попадя как признак "системности".

> The logic of the fiber_cancel should remain the same, excepting checking for systemness.

`fiber_cancel()` можно менять. Его можно заставить игнорить системные файберы, в этом нет проблемы.

> fiber_cancel_system is supposed to work the same way the fiber_cancel currently works

Yes.

> what will be the differences in behavior

`fiber_cancel()` ignores system fibers and it is public, `fiber_cancel_system()` doesn't ignore the system fibers and is private (not in module.h, not in exports file).

In the meantime I found another related bug: https://github.com/tarantool/tarantool/issues/7537. We probably need a more serious solution which would cover all those recent fiber issues.

Reply:
> fiber_cancel() ignores system fibers and it is public, fiber_cancel_system() doesn't ignore the system fibers and is private

Exactly the thing which is done here: https://github.com/tarantool/tarantool/pull/7524. However, `fiber_cancel_system` is added to fiber.h as there's no internal fiber API and, of course, it doesn't cover #7537.

Reply:
The problem is not in `fiber.h`. Only one part of `fiber.h` is public. See the comments `/** \cond public */` and `/** \endcond public */` in the file.

Reply:
> Anyway, I can't merge this PR unitl @Gerold103 approves

Just tell me and I will approve 😉

Reply:
> > Anyway, I can't merge this PR unitl @Gerold103 approves
> 
> Just tell me and I will approve wink

@Gerold103 I think that this ticket does resolve #7473 - it actually prohibits killing system fibers via Lua API. True, we can still kill a system fiber via C API, but I think we should file separate tickets for this and merge this PR as is, because it's a step forward. Please approve.